File: frontend/js/views/free_cash_flow.js
// Free cash flow — FCF, FCF margin, FCF yield, and cash-conversion quality,
// via /calc/free-cash-flow. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 0, maximumFractionDigits: 0 });
const pct = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }) + '%');
const ratio = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 3 }) + '×');
const VIEW = 'free-cash-flow';
let lastReport = null;
let lastBody = null;

export async function renderFreeCashFlow(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.fcf.h1.title">// FREE CASH FLOW</span></h1>
        <p class="muted small" data-i18n="view.fcf.hint.intro">
            The cash a business throws off after maintaining and growing its asset base — what's
            actually available for dividends, buybacks, and debt paydown. FCF = operating cash flow −
            capex. The margin, yield, and FCF-to-net-income (cash-conversion quality, above 1 means
            earnings are backed by cash) round it out. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.fcf.h2.inputs">The company</h2>
            <form id="fcf-form" class="inline-form">
                <label><span data-i18n="view.fcf.label.ocf">Operating cash flow ($)</span>
                    <input type="number" step="0.01" name="operating_cash_flow_usd" value="500" required></label>
                <label><span data-i18n="view.fcf.label.capex">Capital expenditures ($)</span>
                    <input type="number" step="0.01" min="0" name="capital_expenditures_usd" value="150" required></label>
                <label><span data-i18n="view.fcf.label.revenue">Revenue ($)</span>
                    <input type="number" step="0.01" min="0" name="revenue_usd" value="2000"></label>
                <label><span data-i18n="view.fcf.label.mcap">Market cap ($)</span>
                    <input type="number" step="0.01" min="0" name="market_cap_usd" value="5000"></label>
                <label><span data-i18n="view.fcf.label.ni">Net income ($)</span>
                    <input type="number" step="0.01" name="net_income_usd" value="400"></label>
            </form>
            <div id="fcf-tools" class="ce-toolbar"></div>
            <button type="button" id="fcf-sens-btn" class="ce-tool" data-i18n="calc.enh.sens.run">▦ Sensitivity</button>
            <div id="fcf-sens" class="ce-sens"></div>
        </div>
        <div id="fcf-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#fcf-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        return {
            operating_cash_flow_usd: Number(fd.get('operating_cash_flow_usd')) || 0,
            capital_expenditures_usd: Number(fd.get('capital_expenditures_usd')) || 0,
            revenue_usd: Number(fd.get('revenue_usd')) || 0,
            market_cap_usd: Number(fd.get('market_cap_usd')) || 0,
            net_income_usd: Number(fd.get('net_income_usd')) || 0,
        };
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcFreeCashFlow(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.fcf.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#fcf-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'free-cash-flow.csv' });
    mount.querySelector('#fcf-sens-btn').addEventListener('click', () => runSens(mount, readBody(), tok));
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['free_cash_flow_usd', r.free_cash_flow_usd],
        ['fcf_margin_pct', r.fcf_margin_pct],
        ['fcf_yield_pct', r.fcf_yield_pct],
        ['fcf_to_net_income', r.fcf_to_net_income],
    ];
}

function renderResult(mount, r, body) {
    const el = mount.querySelector('#fcf-result');
    const fcfClass = r.free_cash_flow_usd >= 0 ? 'pos' : 'neg';
    const b = body || lastBody || {};
    // FCF = operating cash flow minus capex.
    const chart = enh.svgBarChart([
        { label: 'OCF', value: b.operating_cash_flow_usd || 0 },
        { label: 'Capex', value: -(b.capital_expenditures_usd || 0) },
        { label: 'FCF', value: r.free_cash_flow_usd },
    ]);
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.fcf.h2.result">The cash flow</h2>
            <div class="cards">
                <div class="card ${fcfClass}"><div class="label" data-i18n="view.fcf.card.fcf">Free cash flow</div>
                    <div class="value ${fcfClass}">${money(r.free_cash_flow_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.fcf.card.yield">FCF yield</div>
                    <div class="value">${pct(r.fcf_yield_pct)}</div></div>
                <div class="card"><div class="label" data-i18n="view.fcf.card.margin">FCF margin</div>
                    <div class="value">${pct(r.fcf_margin_pct)}</div></div>
            </div>
            ${chart}
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.fcf.row.fcf">Free cash flow</td><td>${money(r.free_cash_flow_usd)}</td></tr>
                    <tr><td data-i18n="view.fcf.row.margin">FCF margin</td><td>${pct(r.fcf_margin_pct)}</td></tr>
                    <tr><td data-i18n="view.fcf.row.yield">FCF yield</td><td>${pct(r.fcf_yield_pct)}</td></tr>
                    <tr class="emph"><td data-i18n="view.fcf.row.quality">FCF / net income</td><td>${ratio(r.fcf_to_net_income)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

async function runSens(mount, base, tok) {
    const panel = mount.querySelector('#fcf-sens');
    panel.innerHTML = `<p class="muted small" data-i18n="calc.enh.sens.running">Computing…</p>`; applyUiI18n(panel);
    // x: capex 0 -> 2x OCF; y: operating cash flow 0.5x -> 1.5x. Output: free cash flow.
    const ocf = base.operating_cash_flow_usd || 500;
    const xVals = enh.linspace(0, ocf * 2, 5);
    const yVals = enh.linspace(ocf * 0.5, ocf * 1.5, 5);
    const { cells } = await enh.runSensitivity({ base, xKey: 'capital_expenditures_usd', yKey: 'operating_cash_flow_usd', xVals, yVals, compute: (b) => api.calcFreeCashFlow(b), pick: (r) => (r ? r.free_cash_flow_usd : null) });
    if (!viewIsCurrent(tok)) return;
    panel.innerHTML = enh.renderSensitivityTable({ xVals, yVals, cells, fmt: (v) => (v == null ? '—' : '$' + Math.round(v)), xfmt: (v) => '$' + Math.round(v), yfmt: (v) => '$' + Math.round(v), xName: t('view.fcf.label.capex') || 'Capex', yName: t('view.fcf.label.ocf') || 'OCF' });
}

File: crates/traderview-core/src/free_cash_flow.rs
//! Free cash flow and the ratios value investors read off it.
//!
//! ```text
//! FCF            = operating cash flow − capital expenditures
//! FCF margin     = FCF / revenue
//! FCF yield      = FCF / market cap
//! FCF / net income = cash-conversion quality (>1 = earnings backed by cash)
//! ```
//!
//! FCF is the cash a business throws off after maintaining and growing its
//! asset base — what's actually available for dividends, buybacks, and debt
//! paydown.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct FcfInput {
    pub operating_cash_flow_usd: f64,
    pub capital_expenditures_usd: f64,
    #[serde(default)]
    pub revenue_usd: f64,
    #[serde(default)]
    pub market_cap_usd: f64,
    #[serde(default)]
    pub net_income_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct FcfResult {
    /// operating cash flow − capex.
    pub free_cash_flow_usd: f64,
    /// FCF / revenue, percent; `None` if revenue ≤ 0.
    pub fcf_margin_pct: Option<f64>,
    /// FCF / market cap, percent; `None` if market cap ≤ 0.
    pub fcf_yield_pct: Option<f64>,
    /// FCF / net income; `None` if net income ≤ 0. Above 1 means cash exceeds
    /// reported earnings (high quality).
    pub fcf_to_net_income: Option<f64>,
}

pub fn analyze(input: &FcfInput) -> FcfResult {
    let fcf = input.operating_cash_flow_usd - input.capital_expenditures_usd;

    FcfResult {
        free_cash_flow_usd: fcf,
        fcf_margin_pct: if input.revenue_usd > 0.0 {
            Some(fcf / input.revenue_usd * 100.0)
        } else {
            None
        },
        fcf_yield_pct: if input.market_cap_usd > 0.0 {
            Some(fcf / input.market_cap_usd * 100.0)
        } else {
            None
        },
        fcf_to_net_income: if input.net_income_usd > 0.0 {
            Some(fcf / input.net_income_usd)
        } else {
            None
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-6
    }

    fn base() -> FcfInput {
        FcfInput {
            operating_cash_flow_usd: 500.0,
            capital_expenditures_usd: 150.0,
            revenue_usd: 2000.0,
            market_cap_usd: 5000.0,
            net_income_usd: 400.0,
        }
    }

    #[test]
    fn fcf_value() {
        assert!(close(analyze(&base()).free_cash_flow_usd, 350.0));
    }

    #[test]
    fn fcf_margin() {
        assert!(close(analyze(&base()).fcf_margin_pct.unwrap(), 17.5));
    }

    #[test]
    fn fcf_yield() {
        assert!(close(analyze(&base()).fcf_yield_pct.unwrap(), 7.0));
    }

    #[test]
    fn fcf_to_net_income() {
        assert!(close(analyze(&base()).fcf_to_net_income.unwrap(), 0.875));
    }

    #[test]
    fn negative_fcf_when_capex_exceeds_ocf() {
        let mut i = base();
        i.capital_expenditures_usd = 700.0;
        assert!(analyze(&i).free_cash_flow_usd < 0.0);
    }

    #[test]
    fn zero_revenue_guards_margin() {
        let mut i = base();
        i.revenue_usd = 0.0;
        assert!(analyze(&i).fcf_margin_pct.is_none());
    }

    #[test]
    fn zero_market_cap_guards_yield() {
        let mut i = base();
        i.market_cap_usd = 0.0;
        assert!(analyze(&i).fcf_yield_pct.is_none());
    }

    #[test]
    fn high_quality_when_fcf_exceeds_earnings() {
        let mut i = base();
        i.operating_cash_flow_usd = 600.0; // FCF 450 > NI 400
        assert!(analyze(&i).fcf_to_net_income.unwrap() > 1.0);
    }
}
